src/Juvix/Compiler/Tree/EvaluatorEff.hs

```diff
@@ -0,0 +1,414 @@
+        goCase NodeCase {..} = do
+          arg' <- eval' _nodeCaseArg
+          case arg' of
+            v@(ValConstr Constr {..}) ->
+              case find (\CaseBranch {..} -> _caseBranchTag == _constrTag) _nodeCaseBranches of
+                Just CaseBranch {..} -> goCaseBranch v _caseBranchSave _caseBranchBody
+                Nothing -> do
+                  def <- maybe (evalError "no matching branch") return _nodeCaseDefault
+                  goCaseBranch v False def
+            _ ->
+              evalError "expected a constructor"
+
+        withExtendedTemp :: Value -> Eff r' a -> Eff r' a
+        withExtendedTemp v m = do
+          ctx <- E.ask
+          withCtx (over evalCtxTemp (BL.cons v) ctx) m
+
+        goCaseBranch :: Value -> Bool -> Node -> Eff r' Value
+        goCaseBranch v bSave body
+          | bSave = withExtendedTemp v (eval' body)
+          | otherwise = eval' body
+
+        goSave :: NodeSave -> Eff r' Value
+        goSave NodeSave {..} = do
+          v <- eval' _nodeSaveArg
+          withExtendedTemp v (eval' _nodeSaveBody)
+
+printValue :: InfoTable -> Value -> Text
+printValue tab = \case
+  ValString s -> s
+  v -> ppPrint tab v
+
+valueToNode :: Value -> Node
+valueToNode = \case
+  ValInteger i -> mkConst $ ConstInt i
+  ValBool b -> mkConst $ ConstBool b
+  ValString s -> mkConst $ ConstString s
+  ValUnit -> mkConst ConstUnit
+  ValVoid -> mkConst ConstVoid
+  ValConstr Constr {..} ->
+    AllocConstr
+      NodeAllocConstr
+        { _nodeAllocConstrInfo = mempty,
+          _nodeAllocConstrTag = _constrTag,
+          _nodeAllocConstrArgs = map valueToNode _constrArgs
+        }
+  ValClosure Closure {..} ->
+    AllocClosure
+      NodeAllocClosure
+        { _nodeAllocClosureInfo = mempty,
+          _nodeAllocClosureFunSymbol = _closureSymbol,
+          _nodeAllocClosureArgs = map valueToNode _closureArgs
+        }
+
+runError :: Eff (E.Error err ': r) x -> Eff r (Either err x)
+runError = fmap (mapLeft snd) . E.runError
+
+mapError :: (E.Error b :> r) => (a -> b) -> Eff (E.Error a ': r) x -> Eff r x
+mapError f = E.runErrorWith (\_ e -> E.throwError (f e))
+
+hEvalIOEither ::
+  forall m.
+  (MonadIO m) =>
+  Handle ->
+  Handle ->
+  InfoTable ->
+  FunctionInfo ->
+  m (Either TreeError Value)
+hEvalIOEither hin hout infoTable funInfo = do
+  let x :: Eff '[Output Value, E.Error EvalError, E.Error TreeError, IOE] Value
+      x = do
+        v <- eval infoTable (funInfo ^. functionCode)
+        hRunIO hin hout infoTable v
+  let handleTrace :: forall q. (MonadIO q) => Value -> q ()
+      handleTrace = liftIO . hPutStrLn hout . printValue infoTable
+  liftIO
+    . runEff
+    . runError @TreeError
+    . mapError toTreeError
+    . runOutputEff handleTrace
+    $ x
+
+-- | Interpret IO actions.
+hRunIO :: forall r. (IOE :> r, E.Error EvalError :> r, Output Value :> r) => Handle -> Handle -> InfoTable -> Value -> Eff r Value
+hRunIO hin hout infoTable = \case
```

Comment:
Could this be abstracted to work for both the effects and non-effects version? Maybe it's not worth it if we're about to leave one of the versions only in the near future

Reply:
I don't see how we can easily refactor this. Moreover, this is something temporary